phoenix-core-client/src/main/java/org/apache/phoenix/jdbc/PhoenixPrefetchedResultSet.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.phoenix.jdbc;
+
+import java.sql.SQLException;
+import java.util.List;
+
+import org.apache.phoenix.compile.RowProjector;
+import org.apache.phoenix.compile.StatementContext;
+import org.apache.phoenix.schema.tuple.Tuple;
+
+/**
+ * Phoenix ResultSet implementation with prefetched rows. It is expected that the implementation
+ * does not need to use any iterators to make server side RPC call.
+ */
+public class PhoenixPrefetchedResultSet extends PhoenixResultSet {
+
+  private final List<Tuple> prefetchedRows;
+  private int prefetchedRowsIndex;
+
+  public PhoenixPrefetchedResultSet(RowProjector rowProjector,
+      StatementContext ctx, List<Tuple> prefetchedRows) throws SQLException {
+    super(null, rowProjector, ctx);
+    this.prefetchedRows = prefetchedRows;
+    this.prefetchedRowsIndex = 0;
+  }
+
+  @Override
+  protected Tuple getCurrentRowImpl() {
+    return prefetchedRows.get(prefetchedRowsIndex++);
```

Comment:
When a row is deleted, there will only be a single row returned correct? Is there a scenario in which the size of `prefetchedRows` is more than 1?

In any case, JDBC contract for `next()` is to return a `false` when there are no more rows, but looking at the implementation of `getCurrentRow` in the base class, it looks like user wlll see an `IndexOutOfBoundsException`? We would also need test coverage for this.

Reply:
Yes I was planning to return null but thought of avoiding the extra if condition because this implementation is only meant to be used by single row Tuple projection to ResultSet conversion for the purpose of atomic update returning single row. Atomic update/delete can not return more than single row or even status (1 for successful update and 0 for failure to update) for more than single row so the usecase itself is limited. Also, usually the application of prefetched result set should know how many rows they are expecting in advance. In this case, it's only one.

This is good point @haridsv, let me put this comment to make it clear. Avoiding if condition was just meant to make this implementation simpler and clean.

Reply:
I understand the reasoning, but I still think breaking the JDBC contract is not a good idea. May be you can find an alternative to honor the contract without adding any overhead?